src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs

```diff
@@ -64,24 +65,28 @@ public PipeWriter Output
             }
         }
 
-        public override async ValueTask DisposeAsync()
+        public override ValueTask DisposeAsync()
         {
-            if (_disposed)
+            lock (_disposeLock)
             {
-                return;
+                if (_disposed)
+                {
+                    return default;
+                }
+                _disposed = true;
             }
 
             Input.Complete();
             _output.Writer.Complete();
 
-            if (_outputTask != null)
+            if (_outputTask == null || _outputTask.IsCompletedSuccessfully)
             {
                 // Wait for the output task to complete, this ensures that we've copied
```

Comment:
Does this comment still make sense?

Reply:
I'd just move it to the line above `return new ValueTask(_outputTask);`

Reply:
Whoops missed this, I'll do this in a follow-up PR later today.